agent/acs/handler/heartbeat_handler_test.go

```diff
@@ -0,0 +1,95 @@
+// +build unit
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package handler
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecsacs"
+	mock_wsclient "github.com/aws/amazon-ecs-agent/agent/wsclient/mock"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/golang/mock/gomock"
+)
+
+const (
+	heartbeatMessageId = "heartbeatMessageId"
+	heartbeatHealthy   = true
+)
+
+func TestAckHeartbeatMessage(t *testing.T) {
+	heartbeatReceived := &ecsacs.HeartbeatMessage{
+		MessageId: aws.String(heartbeatMessageId),
+		Healthy:   aws.Bool(heartbeatHealthy),
+	}
+
+	heartbeatAckExpected := &ecsacs.HeartbeatAckRequest{
+		MessageId: aws.String(heartbeatMessageId),
+	}
+
+	validateHeartbeatAck(t, heartbeatReceived, heartbeatAckExpected)
+}
+
+func TestAckHeartbeatMessageWithoutMessageId(t *testing.T) {
+	heartbeatReceived := &ecsacs.HeartbeatMessage{
+		Healthy: aws.Bool(heartbeatHealthy),
+	}
+
+	heartbeatAckExpected := &ecsacs.HeartbeatAckRequest{
+		MessageId: nil,
+	}
+
+	validateHeartbeatAck(t, heartbeatReceived, heartbeatAckExpected)
+}
+
+func TestAckHeartbeatMessageEmpty(t *testing.T) {
+	heartbeatReceived := &ecsacs.HeartbeatMessage{}
+
+	heartbeatAckExpected := &ecsacs.HeartbeatAckRequest{
+		MessageId: nil,
+	}
+
+	validateHeartbeatAck(t, heartbeatReceived, heartbeatAckExpected)
+}
+
+func validateHeartbeatAck(t *testing.T, heartbeatReceived *ecsacs.HeartbeatMessage, heartbeatAckExpected *ecsacs.HeartbeatAckRequest) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	var heartbeatAckSent *ecsacs.HeartbeatAckRequest
+
+	mockWsClient := mock_wsclient.NewMockClientServer(ctrl)
+	mockWsClient.EXPECT().MakeRequest(gomock.Any()).Do(func(message *ecsacs.HeartbeatAckRequest) {
+		heartbeatAckSent = message
+		cancel()
+	}).Times(1)
+
+	handler := newHeartbeatHandler(ctx, mockWsClient)
+	go handler.sendHeartbeatAck()
+
+	handler.handleSingleHeartbeatMessage(heartbeatReceived)
+
+	// wait till we get an ack from heartbeatAckMessageBuffer
+	select {
+	case <-ctx.Done():
```

Comment:
nit: select is not necessary here, this can just be `<-ctx.Done()`

Reply:
Fixed in next rev